### PR TITLE
Upgrade colored and prepare for 0.7.0 release

### LINKED ISSUE #113: Getting rid of `atty` as a transitive dependency (via `colored`)
So. `atty` has an unsoundness issue ([RUSTSEC-2021-0145](https://rustsec.org/advisories/RUSTSEC-2021-0145.html)), seems to be 100% unmaintained and people are moving away from it in general (https://github.com/clap-rs/clap/pull/4249, https://github.com/rust-cli/env_logger/pull/248, ...). `atty` is part of the dependency tree for `fern` via its direct dependency on `colored`. However, `colored` also does not seem to be too quick in wanting to fix the `atty` issues (https://github.com/mackwic/colored/pull/122).

So I'm posting this issue here also, in order to track the possibility of depending on `fern` without pulling in an unsoundness issue. For one it can act as some pressure on `colored` to finally merge and release that fix. Or `fern` can consider changing library for colors.

### PR DESCRIPTION
**Pull request by faern**
Fixes #113

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,14 @@
 Unreleased
 ==========
 
+
+0.7.0 (2023-07-12)
```
Comment (faern) [inline CHANGELOG.md:5]: The date might need to be adjusted unless this PR is accepted quickly. Or the package maintainers might want to handle this part of releasing themselves? Just tell me how you want it. I went with the path where this PR does everything needed before `cargo publish` for now, to make it easy :)

**Comment by faern** [comment]
`time` bumped the MSRV to 1.65.0 in version `0.3.21`. They (correctly) did not release this as a breaking change. But this also means that it's kind of pointless to track MSRV at all. Because we can't guarantee anything about our MSRVs when we can't do the same for our dependencies.

What we can do is to run the CI on `cargo +nightly update -Z minimal-versions && cargo test` to verify that our MSRV at least holds for the *oldest* compatible versions of all our dependencies.. Or skip the MSRV CIs completely. The maintainers need to step in here. But this issue has nothing with `colored` to do, it's just `syslog`.

**Comment by LingMan** [comment]
Other options would be to run the MSRV CI with `cargo update -p time --precise 0.3.20`, which would only downgrade that specific crate, or committing a locally crated `Cargo.lock.msrv` and then doing `cp Cargo.lock.msrv Cargo.lock`.
Both of these would avoid the need to install nightly rust on the MSRV CI job.

File: src/lib.rs [hunk 2]
```diff
@@ -1,43 +1,7 @@
 #![deny(missing_docs)]
-#![doc(html_root_url = "https://docs.rs/fern/0.6.2")]
+#![doc(html_root_url = "https://docs.rs/fern/0.7.0")]
```
[the remark this answers is not in the document]
Comment (LingMan) [reply src/lib.rs:2]: Crates.io needs an external request to show the docs.rs link if it's not given in the Cargo.toml. It didn't show up for me for the longest time since my uMatrix blocked said request. Arguably my fault but you can still see it taking a split second longer to show up even when it works as expected.
There's currently no version number in the `documentation` link (nor should there be one), so no need to update it.
IMO it should stay as it doesn't really require maintenance.

**Comment by polarathene** [comment]
> this also means that it's kind of pointless to track MSRV at all. Because we can't guarantee anything about our MSRVs when we can't do the same for our dependencies.

> IMHO that's a very temporary and fragile solution. It's just a matter of time before some other dependency bumps their MSRV and this CI job fails for no apparent reason again.

You can use `-Z msrv-policy` with nightly toolchain to resolve crates to an MSRV if they properly maintain their `rust-version` in `Cargo.toml`.

Useful for generating a `Cargo.lock.msrv` for CI usage, where you may have a few crates that need version pinning if they're not compatible yet with `-Z msrv-policy`.

---

Just to clarify, a project can set an MSRV and ensure that it satisfies that MSRV via CI, but this does not require the ability to have the MSRV supported without version pinning. It's the same as what the user would need to do to build with the MSRV, bumping it to avoid any pinning is just for convenience.

`rust-version` in `Cargo.toml` should convey what  the actual MSRV is to build without failure (_because the project itself uses an API or a [dependency that actually requires an MSRV bump to support the functionality](https://github.com/daboross/fern/issues/113#issuecomment-1616729195)_). If you're just interested in what the MSRV is without a `Cargo.lock` use something like `cargo-msrv` in CI to track that, it's a different expectation.

Upstream dependencies can MSRV bump as non-breaking semver change without issue, especially when they leverage `rust-version`. [This explains it well](https://github.com/sgodwincs/ordered-multimap-rs/issues/26#issuecomment-1820315686).

`config-rs 0.13.x` still supports a MSRV of `1.56` by using [`-Z msrv-policy` and only three pins](https://github.com/mehcode/config-rs/pull/495).

---

## TL;DR

`rust-version` is where the value is at for downstreams, provided it's maintained correctly (_upstreams have no impact on MSRV then when they also maintain their `rust-version`_).

---

This PR has been left open for a while, and no new commits to `main` since March. Perhaps @daboross could add some additional maintainers?

**Comment by pronebird** [comment]
Bump

**Comment by faern** [comment]
The original reason for this PR is gone (getting rid of `atty` as a transitive dependency). But I still think it makes sense to do this upgrade to `colored 2`. This gets rid of the also stale `winapi` crate, and in general follow the ecosystem forward. Sadly this PR also has to deal with MSRV stuff. Not really because this PR changes much around it, but because the CI for `fern` is currently broken if running straight off of `main`.

I went with the solution to check in a `Cargo.lock.msrv` in order to be able to maintain CI MSRV checks. However, I had to bump the lower bound on `rust-version` anyway, since the lockfile uses the new format introduced in Rust 1.38. Maybe it's possible to craft a lockfile of the older format :shrug: but to me that seems like madness for such an extreme backwards compatibility.

**Comment by daboross** [comment]
Thank you for making this! And sorry to leave it completely unattended for ~15 months.

@polarathene you're correct that I've left this crate unmaintained - not an ideal state for something people depend on. I think `fern` could potentially use someone else who could merge small-but-important things like this, but given the overall maintenance burden of this crate is pretty small, I don't think it's that urgent.

In any case, I'll be making a new `0.x` revision with the new `colored` crate now. I'll base it on this PR - we'll see if GitHub considers it merged or not :shrug:.

In the past I've just upped the MSRV to whatever `log` uses, given that's our only non-optional dependency. I Especially since the upcoming version will be a major revision, I think it's imminently reasonable to just do that and jump to 1.60.0 - I'll be doing that.

**Comment by polarathene** [comment]
> @polarathene you're correct that I've left this crate unmaintained - not an ideal state for something people depend on. I think `fern` could potentially use someone else who could merge small-but-important things like this.

That's ok, it happens. See `config-rs` in a similar bind where they're working towards moving the repo to `rust-cli` organization for maintenance.

I think you could probably engage with `rust-cli` as well if you like? From what I understand you'd still have the ability to maintain the repo, but as part of the `rust-cli` organization they'll ensure other maintainers are available should you be lacking the spare time :)
